File: src/bj12865/Main.java
package bj12865;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] testCase=br.readLine().split(" ");
        int count= Integer.parseInt(testCase[0]);
        int weight= Integer.parseInt(testCase[1]);
        int[][] dp=new int[count][weight+1];
        for(int i=0; i<count; i++){
            String[] read=br.readLine().split(" ");
            int w= Integer.parseInt(read[0]);
            int v= Integer.parseInt(read[1]);
            for(int j=1; j<=weight; j++){
                if(i==0&&j>=w){
                    dp[i][j]=v;
                } else if(i>0){
                    if(j-w>=0)
                        dp[i][j]=Math.max(v+dp[i-1][j-w], dp[i-1][j]);
                    else
                        dp[i][j]=dp[i-1][j];
                }
            }
        }
        System.out.println(dp[count-1][weight]);
    }
}

File: kotlin/src/main/kotlin/bj12865/Main.kt
package bj12865

import java.io.*
import kotlin.math.max

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (count, weight) = br.readLine().split(" ").map { it.toInt() }
    val dp = Array(count) { IntArray(weight + 1) }
    for (i in 0 until count) {
        val (w, v) = br.readLine().split(" ").map { it.toInt() }
        for (j in 1..weight) {
            if (i == 0 && j >= w) {
                dp[i][j] = v
            } else if (i > 0) {
                if (j >= w) {
                    dp[i][j] = max(v + dp[i - 1][j - w], dp[i - 1][j])
                } else {
                    dp[i][j] = dp[i - 1][j]
                }
            }
        }
    }
    println(dp[count - 1][weight])
}
